File: Fort.ind UWP/Models/UserProfile.vb
Imports System.Runtime.Serialization

''' <summary>
''' Represents a user profile with account information
''' Designed for local storage now, ready for server sync later
''' </summary>
<DataContract>
Public Class UserProfile

    ''' <summary>
    ''' Unique identifier for the user (GUID for local, can map to server ID later)
    ''' </summary>
    <DataMember>
    Public Property UserId As String

    ''' <summary>
    ''' Display name shown in the app
    ''' </summary>
    <DataMember>
    Public Property DisplayName As String

    ''' <summary>
    ''' Username for login
    ''' </summary>
    <DataMember>
    Public Property Username As String

    ''' <summary>
    ''' Hashed password (for local auth, will use tokens for server)
    ''' </summary>
    <DataMember>
    Public Property PasswordHash As String

    ''' <summary>
    ''' User's email address
    ''' </summary>
    <DataMember>
    Public Property Email As String

    ''' <summary>
    ''' Optional bio/description
    ''' </summary>
    <DataMember>
    Public Property Bio As String

    ''' <summary>
    ''' Path to profile picture (local path or URL)
    ''' </summary>
    <DataMember>
    Public Property ProfilePicturePath As String

    ''' <summary>
    ''' When the account was created
    ''' </summary>
    <DataMember>
    Public Property CreatedDate As DateTime

    ''' <summary>
    ''' Last login timestamp
    ''' </summary>
    <DataMember>
    Public Property LastLoginDate As DateTime

    ''' <summary>
    ''' User preferences and settings
    ''' </summary>
    <DataMember>
    Public Property Preferences As UserPreferences

    ''' <summary>
    ''' Creates a new empty profile
    ''' </summary>
    Public Sub New()
        UserId = Guid.NewGuid().ToString()
        CreatedDate = DateTime.Now
        Preferences = New UserPreferences()
    End Sub

    ''' <summary>
    ''' Creates a new profile with basic info
    ''' </summary>
    Public Sub New(username As String, displayName As String, email As String)
        Me.New()
        Me.Username = username
        Me.DisplayName = displayName
        Me.Email = email
    End Sub

End Class

''' <summary>
''' User preferences and settings
''' </summary>
<DataContract>
Public Class UserPreferences

    ''' <summary>
    ''' Enable Live Tile updates
    ''' </summary>
    <DataMember>
    Public Property EnableLiveTile As Boolean = True

    ''' <summary>
    ''' Enable notifications
    ''' </summary>
    <DataMember>
    Public Property EnableNotifications As Boolean = True

    ''' <summary>
    ''' Theme preference (Dark, Light, System)
    ''' </summary>
    <DataMember>
    Public Property Theme As String = "Dark"

    ''' <summary>
    ''' Remember login between sessions
    ''' </summary>
    <DataMember>
    Public Property RememberLogin As Boolean = True

End Class

File: Fort.ind UWP/Models/UserProfile.cs
using System;
using System.Runtime.Serialization;

namespace Fort.ind_UWP
{
    /// <summary>
    /// A cached snapshot of the signed-in fort.social (Misskey) account.
    /// The instance is the source of truth; this is a local copy for offline display.
    /// </summary>
    [DataContract]
    public class UserProfile
    {

        /// <summary>
        /// The Misskey user ID on the instance.
        /// </summary>
        [DataMember]
        public string UserId { get; set; }

        /// <summary>
        /// Misskey username (without the leading @ or host).
        /// </summary>
        [DataMember]
        public string Username { get; set; }

        /// <summary>
        /// Remote instance host, or null/empty for a local fort.social account.
        /// </summary>
        [DataMember]
        public string Host { get; set; }

        /// <summary>
        /// Display name shown in the app
        /// </summary>
        [DataMember]
        public string DisplayName { get; set; }

        /// <summary>
        /// Bio/description, as set on fort.social
        /// </summary>
        [DataMember]
        public string Bio { get; set; }

        /// <summary>
        /// URL of the user's avatar image on the instance.
        /// </summary>
        [DataMember]
        public string AvatarUrl { get; set; }

        /// <summary>
        /// When the fort.social account was created
        /// </summary>
        [DataMember]
        public DateTime CreatedDate { get; set; }

        /// <summary>
        /// Last time this app signed the user in
        /// </summary>
        [DataMember]
        public DateTime LastLoginDate { get; set; }

        /// <summary>
        /// User preferences and settings.
        /// Getter ensures this is never null even after deserialization
        /// (DataContractJsonSerializer bypasses the constructor).
        /// </summary>
        [DataMember]
        public UserPreferences Preferences
        {
            get
            {
                if (_preferences == null) _preferences = new UserPreferences();
                return _preferences;
            }
            set
            {
                _preferences = value;
            }
        }
        private UserPreferences _preferences;

        public UserProfile()
        {
            Preferences = new UserPreferences();
        }

        /// <summary>
        /// Returns a deep copy of this profile. Used so edits can be saved on a detached copy
        /// and only swapped into the shared CurrentUser after the save succeeds, avoiding
        /// readers observing a half-applied mutation.
        /// </summary>
        public UserProfile Clone()
        {
            UserProfile copy = new UserProfile()
            {
                UserId = this.UserId,
                Username = this.Username,
                Host = this.Host,
                DisplayName = this.DisplayName,
                Bio = this.Bio,
                AvatarUrl = this.AvatarUrl,
                CreatedDate = this.CreatedDate,
                LastLoginDate = this.LastLoginDate
            };
            var prefs = this.Preferences;
            copy.Preferences = new UserPreferences()
            {
                EnableLiveTile = prefs.EnableLiveTile,
                EnableNotifications = prefs.EnableNotifications,
                Theme = prefs.Theme
            };
            return copy;
        }

    }

    /// <summary>
    /// User preferences and settings
    /// </summary>
    [DataContract]
    public class UserPreferences
    {

        /// <summary>
        /// Enable Live Tile updates
        /// </summary>
        [DataMember]
        public bool EnableLiveTile { get; set; } = true;

        /// <summary>
        /// Enable notifications
        /// </summary>
        [DataMember]
        public bool EnableNotifications { get; set; } = true;

        /// <summary>
        /// Theme preference (Dark, Light, System)
        /// </summary>
        [DataMember]
        public string Theme { get; set; } = "Dark";

    }
}
